File: apps/api/apps/librarian/routes/borrowers.js
const { fetchBorrowers, fetchBorrower, updateBorrower } = require('../../../services/borrowers');

const express = require('express');
const router = express.Router();

router.get('/', async (req, res) => {
    res.send(await fetchBorrowers());
});

router.get('/:id', async (req, res) => {
    res.send(await fetchBorrower({ id: req.params.id }));
});

router.patch('/:id/contact', async (req, res) => {
    try {
        const { id } = req.params;
        const { email, phone } = req.body;

        await updateBorrower(id, { email, phone });
        res.status(204).send();
    } catch (error) {
        console.error(error);
        res.status(error.statusCode || 500).send({ errors: [error] });
    }
});

module.exports = router;

File: apps/librarian/lib/api/borrowers.dart
part of 'api.dart';

Future<Response> fetchBorrower(String id) async {
  return await DioClient.instance.get('/borrowers/$id');
}

Future<Response> fetchBorrowers() async {
  return await DioClient.instance.get('/borrowers');
}

Future<Response> updateBorrower(
  String id, {
  String? email,
  String? phone,
}) async {
  dynamic data = {};

  if (email != null) {
    data['email'] = email;
  }

  if (phone != null) {
    data['phone'] = phone;
  }

  return await DioClient.instance.patch('/borrowers/$id/contact', data: data);
}
